ister.py

```diff
@@ -124,26 +130,41 @@ def stream_fetcher(info, streamid):
 def wait_for_process(proc, log_output, show_output):
     """Wait for process 'proc' to finish."""
 
-    # The following code makes an assumption that 'cmd' outputs to 'stdout', but
-    # uses 'stderr' only in case of failure. Therefore we first exhaust the
-    # 'stdout' pipe and then look to the 'stderr' pipe. If 'cmd' outputs to
-    # 'stderr' as part of its normal operation, the below code may cause a
-    # deadlock:
-    # * 'cmd' prints to 'stderr'
-    # * 'stderr' pipe is full and 'cmd' gets blocked
-    # * we keep waiting on 'stdout'
-    # * deadlock
-    output = []
-    for stream in (proc.stdout, proc.stderr):
-        lines = []
-        output.append(lines)
-        for line in stream:
-            decoded_line = line.decode('ascii', 'ignore').rstrip()
-            lines.append(decoded_line)
-            if show_output:
-                LOG.info(decoded_line)
-            elif log_output:
-                LOG.debug(decoded_line)
+    info = {"streams" : (proc.stdout, proc.stderr),
+            "die_now" : False,
+            "queue" : queue.Queue()}
+
+    # Start the stream fetcher threads. They will read the output of the process
+    # and put to the queue.
+    threads = []
+    for streamid in (0, 1):
+        if info["streams"][streamid]:
+            threads.append(threading.Thread(target=stream_fetcher,
+                                            name='cmd-stream-fetcher',
+                                            args=(info, streamid)))
+            threads[-1].start()
+
+    output = ([], [])
+    try:
+        while True:
+            streamid, line = info["queue"].get()
+            if line is not None:
+                # Drop single newline.
+                line = line[:-1]
```

Comment:
So instead of stripping you are only removing the newline which changes the output again I'm not super worried about it but extra whitespace seems odd to be adding to logs.

Reply:
Also this will drop the last character of the partial line added at the end. I think using strip here is probably best.

Reply:
I had a strip and then removed it because I believe it removes more than needed. I'll re-examine.

Reply:
OK, the problem was that extract_full_lines returned all the full lines as one string. Not as a list of lines. So it could return "text\n\n\n", for example, because the command printed three newlines in a row. Using .strip() would remove all the three, and subsequent 'LOG.debug()' would add one. So the end result is that newlines that must have been there would be removed.

I've changed this and now extract_full_lines returns a per-line list, already stripped, so nothing else should be done to the output.